src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp

```diff
@@ -82,21 +82,45 @@ void GenericPlatformManagerImpl_POSIX<ImplClass>::_LockChipStack()
 {
     int err = pthread_mutex_lock(&mChipStackLock);
     assert(err == 0);
+
+#if defined(CHIP_STACK_LOCK_TRACKING_ENABLED)
+    mChipStackIsLocked        = true;
+    mChipStackLockOwnerThread = pthread_self();
+#endif
 }
 
 template <class ImplClass>
 bool GenericPlatformManagerImpl_POSIX<ImplClass>::_TryLockChipStack()
 {
-    return pthread_mutex_trylock(&mChipStackLock) == 0;
+    bool locked = (pthread_mutex_trylock(&mChipStackLock) == 0);
+#if defined(CHIP_STACK_LOCK_TRACKING_ENABLED)
+    if (locked)
+    {
+        mChipStackIsLocked        = true;
+        mChipStackLockOwnerThread = pthread_self();
+    }
+#endif
+    return locked;
 }
 
 template <class ImplClass>
 void GenericPlatformManagerImpl_POSIX<ImplClass>::_UnlockChipStack()
 {
     int err = pthread_mutex_unlock(&mChipStackLock);
     assert(err == 0);
+#if defined(CHIP_STACK_LOCK_TRACKING_ENABLED)
+    mChipStackIsLocked = false;
+#endif
 }
 
+#if defined(CHIP_STACK_LOCK_TRACKING_ENABLED)
+template <class ImplClass>
+bool GenericPlatformManagerImpl_POSIX<ImplClass>::_IsChipStackLockedByCurrentThread() const
+{
+    return !mMainLoopStarted || (mChipStackIsLocked && (pthread_equal(pthread_self(), mChipStackLockOwnerThread)));
```

Comment:
Does this means the CHIP stack can not be locked by current thread if CHIP thread is not started?

Reply:
Should _IsChipStackLockedByCurrentThread be only called within mainloop? This lock is to lock the whole chip stack, I was wondering if we should allow to lock chip stack from another thread other than mainloop.

Reply:
The lock is only useful as a means to manage access from multiple threads.  So yes, other threads must be permitted to lock.  Any code that calls a public interface that requires protection must be able to acquire the lock.

Reply:
@msandstedt My question is if we call this check from a thread other than chip main thread, it always return 'true', does not matter if the CHIP stack has been locked, or the CHIP stack was locked by a different thread. Is this behavior  expected?

Reply:
@msandstedt I think there might be misunderstand of my question. 

Current design is your following statement is only correct when CHIP main thread has already started (mMainLoopStarted = true)

"If the current thread has locked the stack, this will reliably return true. If the current thread has not locked the stack and another thread has, this will semi-reliably return false."

If CHIP main thread has not started (mMainLoopStarted = false), then IsChipStackLockedByCurrentThread always return true regardless, and your statement above is not correct in this case.